plugins/scaffolder/src/components/fields/EntityPicker/EntityPicker.tsx

```diff
@@ -81,38 +79,40 @@ export const EntityPicker = (props: EntityPickerProps) => {
   const getLabel = useCallback(
     (ref: string) => {
       try {
-        return humanizeEntityRef(
-          parseEntityRef(ref, { defaultKind, defaultNamespace }),
-          {
-            defaultKind,
-            defaultNamespace,
-          },
-        );
+        return entityPresentationApi.forEntity(ref, {
+          defaultKind,
+          defaultNamespace,
+        }).snapshot.primaryTitle;
```

Comment:
Here's where you hit the quirks of the initial presentation API design. As you see it's a combo of a snapshot and an Observable. The reason for this is basically React. The snapshot **is synchronous** (not a promise) which is key because we want to avoid blocking the rendering loop. But we also want the presentation API to be able to be asynchronously promise based under the hood because it may need to make service requests to the catalog. So what the [useEntityPresentation hook](https://github.com/backstage/backstage/blob/2ca9dc00b2a1d26cbf8f4700463bdeeb962fec55/plugins/catalog-react/src/apis/EntityPresentationApi/useEntityPresentation.ts#L43) does is to first render whatever the snapshot says and then subscribe to the observable for updates for the REAL end result value. All of this needs to stay "reactive" (cause rerenders in React).

The key here is that the snapshot may be _wildly wrong_ (but we don't know upfront), and the event that comes out of the observable may be the right thing to show.

But the interface is hard to consume for "plain javascript use cases" like the one we're looking at here.

What you would want, instead, is to be able to do `await entityPresentationApi.forEntity(ref).promise` inside the `useAsync` above. Which means you'll have to add that `promise` getter to the presentation API!

Are you up for that? If that sounds too complex, maybe I can help guide.

Reply:
I tried to dig into this but I have a few questions.   
With the changes I made in my latests commits (which mostly rollbacks usage of presentation api to `stringifyEntityRef` and `parseEntityRef`) I only have two reference to the presentation api now. 

One here in the `getLabel` and one in `getOptionLabel` on the Autocomplete. Both needs to get the `primaryTitle` of the presentation api. For this use case, in this context is it really useful to add this `.promise` to the presentation api ? Isn't there another way (like with `stringifyEntityRef` and `parseEntityRef`) to get that primary title ? 

If not possible, is there any case where we will send an outdated value to the `.forEntity` ? From what I see there is a variable called `needsLoad` in `#getEntityForInitialRender` which should be always false if we send `Entity` objects (that we got from the catalogApi, so up to date ?)

I'm a bit confused with this but my ultimate goal now is to retrieve the primary title of each entity